app/src/main/java/com/google/samples/apps/sunflower/compose/plantdetail/PlantDetailScroller.kt

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.samples.apps.sunflower.compose.plantdetail
+
+import androidx.animation.FloatPropKey
+import androidx.animation.Spring
+import androidx.animation.transitionDefinition
+import androidx.compose.Composable
+import androidx.ui.foundation.ScrollerPosition
+import androidx.ui.unit.Density
+import androidx.ui.unit.Dp
+import androidx.ui.unit.dp
+
+private val HeaderTransitionOffset = 32.dp
+private const val ParallaxFactor = 2f
+
+/**
+ * Class that contains derived state for when the toolbar should be shown
+ */
+data class PlantDetailsScroller(
+    val scrollerPosition: ScrollerPosition,
+    val namePosition: Float
+) {
+    fun getToolbarState(density: Density): ToolbarState {
+        return if (namePosition != 0f &&
+            scrollerPosition.value > (namePosition + getTransitionOffset(density))
+        ) {
+            ToolbarState.SHOWN
+        } else {
+            ToolbarState.HIDDEN
+        }
+    }
+
+    private fun getTransitionOffset(density: Density): Float = with(density) {
```

Comment:
What about moving this to a param so that the Dp value is converted to Px once when this is created rather than on every scroll?

Reply:
The caller wraps `PlantDetailsScroller` into a `state`. Thus, it's not possible to pass in `DensityAmbient.current` at construction time.